### ISSUE TITLE
_FileFormat should self-validate on construction

### ISSUE THREAD

**Issue by thermokarst** [qiime2/qiime2#490]
**Improvement Description**
It is currently possible to create invalid file formats via the format constructor. This isn't generally a problem, since users typically "import" their data, which _does_ call the format validator. One common place that formats are instantiated via the constructor though is in unit tests, and as such, if the caller forgets to call `.validate()` on the instance, they can potentially be working with "invalid" data.

**Current Behavior**
The `_FileFormat` constructor knows nothing about `.validate()`.

**Proposed Behavior**
The `_FileFormat`  constructor calls `.validate()` as part of its set up.

**Questions**
1. I suspect we left this off of the initial validate API because we were hesitant to start validating data willy-nilly, before understanding the performance implications. We have slowly been filling in some [validation gaps](https://github.com/qiime2/qiime2/pull/461), but I wonder if I am forgetting something else about this.

**Comment by ebolyen** [comment]
One thing that springs to mind is transformers often initialize an empty version of these in order to create a filesystem reference. Then they are filled in. We would need to preserve that functionality without adversely impacting the API too much

**Comment by Oddant1** [comment]
Expanding on what @ebolyen said, `_FileFormat` doesn't have a constructor, but `FormatBase` does, and it only does some preliminary setup in a similar vein to the transformers. This makes it so dropping a call to `validate` in the `FormatBase` constructor doesn't work because basically nothing about the object has actually been created yet, so the `validate` check inevitably fails immediately. I get the feeling that with the way things are currently set up we would probably need to drop `validate` calls into a significant number of different constructors.

**Comment by Oddant1** [comment]
@ebolyen and I did some experimentation. If you only validate files that were opened in read mode in init then in theory the file should validate because the underlying structure should have already been created. This should catch cases like the one that seems to have initially prompted the creation of this issue in qiime2/q2-demux#104

**Comment by thermokarst** [comment]
@Oddant1, that sounds to me like that should work well! Curious to hear if @ebolyen sees any flaws with that approach.

**Comment by ebolyen** [comment]
I think this is the most likely approach to work, we'll want to look at the lifecycle of these objects carefully though. For instance, we don't want transformers to validate max _every_ time, just sometimes. Similarly, in situations with test data, we probably do want the max validation. 

Overall, it seems complicated, so we'll want to dot our i's and cross our t's. I think this problem is mostly analysis at this point. So I'd be interested in a high level summary from @Oddant1 about what the impacts are and how to mitigate them.

**Comment by Oddant1** [comment]
Yeah, if we just validated in `FormatBase.__init__` every time we construct in 'r' mode, we'll run into a case where pretty much every transformer is validating it's outputs two times. The problem is, I'm not sure if there's currently any way to distinguish between the object being created in a transformer and the object not being created in a transformer. We could pass some sort of argument to the constructor telling it if it's being created in a transformer or not, but then we'd have to retroactively add it to a whole lot of existing initializations if we wanted to do things properly.
